HARK/ConsumptionSaving/ConsPortfolioFrameModel.py

```diff
@@ -0,0 +1,286 @@
+solving, and simulating agents who must allocate their resources
+among consumption, saving in a risk-free asset (with a low return),
+and saving in a risky asset (with higher average return).
+
+This file also demonstrates a "frame" model architecture.
+"""
+import numpy as np
+from scipy.optimize import minimize_scalar
+from copy import deepcopy
+from HARK import NullFunc, Frame, FrameAgentType  # Basic HARK features
+from HARK.ConsumptionSaving.ConsIndShockModel import (
+    IndShockConsumerType,  # PortfolioConsumerType inherits from it
+    utility,  # CRRA utility function
+    utility_inv,  # Inverse CRRA utility function
+    utilityP,  # CRRA marginal utility function
+    utility_invP,  # Derivative of inverse CRRA utility function
+    utilityP_inv,  # Inverse CRRA marginal utility function
+    init_idiosyncratic_shocks,  # Baseline dictionary to build on
+)
+from HARK.ConsumptionSaving.ConsRiskyAssetModel import (
+    RiskyAssetConsumerType
+)
+from HARK.ConsumptionSaving.ConsPortfolioModel import (
+    init_portfolio,
+    solveConsPortfolio,
+    PortfolioConsumerType,
+    PortfolioSolution
+)
+
+from HARK.distribution import combine_indep_dstns, add_discrete_outcome_constant_mean
+from HARK.distribution import (
+    IndexDistribution,
+    Lognormal,
+    MeanOneLogNormal,
+    Bernoulli  # Random draws for simulating agents
+)
+from HARK.interpolation import (
+    
+    LinearInterp,  # Piecewise linear interpolation
+    CubicInterp,  # Piecewise cubic interpolation
+    LinearInterpOnInterp1D,  # Interpolator over 1D interpolations
+    BilinearInterp,  # 2D interpolator
+    ConstantFunction,  # Interpolator-like class that returns constant value
+    IdentityFunction,  # Interpolator-like class that returns one of its arguments
+    ValueFuncCRRA,
+    MargValueFuncCRRA,
+    MargMargValueFuncCRRA
+)
+
+class PortfolioConsumerFrameType(FrameAgentType, PortfolioConsumerType):
+    """
+    A consumer type with a portfolio choice, using Frame architecture.
+
+    A subclass of PortfolioConsumerType for now.
+    This is mainly to keep the _solver_ logic intact.
+    """
+
+    def __init__(self, **kwds):
+        params = init_portfolio.copy()
+        params.update(kwds)
+        kwds = params
+
+        # Initialize a basic consumer type
+        PortfolioConsumerType.__init__(
+            self, **kwds
+        )
+
+        # Set the solver for the portfolio model, and update various constructed attributes
+        self.solve_one_period = solveConsPortfolio
+        self.update()
+
+        ## TODO: Should be defined in the configuration.
+        self.aggs = {'PermShkAggNow' : None, 'PlvlAgg' : None} # aggregate values
+                      # -- handled differently because only one value each per AgentType
+        self.shocks = {'Adjust' : None, 'PermShk' : None, 'Risky': None, 'TranShk' : None}
+        self.controls = {'cNrm' : None, 'Share' : None}
+        self.state_now = {
```

Comment:
was the `_now` not out of fashion now?

Reply:
These are doing some logical work as per the current HARK simulation code.
There, there's both state_now and state_prev, because transition functions can depend on the previous period's state.

I can see how we could do an implementation where the state memory array was duplicated each period, and updated frame-by-frame.

That's an implementation detail, as far as I'm concerned, not a matter of model logic. I'm not opposed to making the implementation consistent for shocks, controls, and states, as another PR. But there may be benefits to keeping state_prev.